### ISSUE TITLE
MapInitializable initialiser problems with new Reflection library dependency.

### ISSUE THREAD

**Issue by sportlabsMike** [GraphQLSwift/GraphQL#37]
Hi,

Since the change to https://github.com/wickwirew/Runtime.git for reflection there's a problem with the default `MapInitializable` initialiser.

In the call `self = try createInstance()`, the new reflection library tries to eagerly create an instance and fill in all the properties with 'default' values. However it doesn't support `enums`, so in the case where the type contains an Enum type property it blows up.

The old implementation of the `MapInitializable` initialiser didn't suffer form this since the `construct` reflection method it used took a `(Property.Description) throws -> Any` parameter that was used to supply all the values as it build the instance (thus it never attempted to create default ones itself)

This means any `InputType`s (or `Arguments` when using Graphiti) that contain enum properties are now broken.

Any ideas for a solution much appreciated...

**Comment by jseibert** [comment]
See my fork of Runtime here: https://github.com/jseibert/Runtime
And my open PR here: https://github.com/wickwirew/Runtime/pull/14

If you can help push the PR along that would be great, I'm short on time at the moment. Otherwise we can officially fork the library and change the dependency.

**Comment by jseibert** [comment]
For reference, I've also forked Graphiti and GraphQL and shifted them to Swift-NIO implementations on top of Vapor 3. If you're interested, you can check them out here. Would love to also get these into upstream somehow but I realize they are pretty specific customizations of the library. 

https://github.com/jseibert/Graphiti
https://github.com/jseibert/GraphQL

**Comment by fruitcoder** [comment]
Awesome @jseibert, this basically solves my issue #36. I didn't realize this is a general graphQL package so it might be wrong to expect the implementation to change to support Vapor. 
Can you explain the difference between Graphiti and GraphQL because I don't seem to get it. I'll test your GraphQL fork tomorrow :)

**Comment by paulofaria** [comment]
@jseibert I see vapor/core uses swift-nio under the covers. Is it possible to strip the dependencies to go for `EventLoopPromise` and `EventLoopFuture` directly? We could provide extensions in a separate repo which would then wrap those basic constructs to higher level vapor types if that helps in any way. What do you think?

And talking about reflection.. I'm thinking *maybe* we could get rid of it in favor of `Codable`? Instead of taking `MapInitializable` we'd take `Decodable`, etc. @sportlabsMike I'm curious about your opinion on this as well.

**Comment by jseibert** [comment]
@paulofaria yes it is, I did that here: https://github.com/jseibert/GraphQL/tree/nio Should be totally doable but I hit a runtime crash and had to move on to other things. Would be great if you had time to wrap it up, should be ~95% good. 

IDEALLY, `Map` goes out the window entirely and we follow Vapor's lead of moving to `Codable`. They even have a well-functioning Decodable-based reflection library (https://github.com/vapor/core/tree/master/Sources/Core/CodableReflection).

**Comment by jseibert** [comment]
Appears to happen on basically any query:

![image](https://user-images.githubusercontent.com/80599/38405156-572aea8e-3923-11e8-84a3-45a19d2895c4.png)

![image](https://user-images.githubusercontent.com/80599/38405228-bc06b1fe-3923-11e8-800d-c41916db9615.png)

**Comment by jseibert** [comment]
A few quick notes for you:
- In the interest of time, I overloaded the Context var to required it be castable as an EventLoopGroup. You'd almost certainly want to make this a more intentional part of the library if you want to take the NIO dependency, and the ELG needs to be passed all the way in from Graphiti.Schema.execute.
- Vapor 3 Core is simply typealias'ed on top of NIO, so there won't be any shims needed to support it out of the box. 
- That said, in a ideal world there would be a shim (or a Graphiti-Provider in Vapor parlance) that can respond directly to a Vapor.Request, which would mean internalizing the parameter ingestion, such as:
```
struct GQL: Decodable {
    let query: String
    let operationName: String?
    let variables: [String: Map]?
}
```

**Comment by sportlabsMike** [comment]
Hi everyone, thanks for the help and great work!

@jseibert I've commented on your PR https://github.com/wickwirew/Runtime/pull/14 so hopefully it will get merged. If it doesn't soon we could consider forking. (In the meantime I'll use your fork to keep me moving, thanks)

@paulofaria I'm in favour of moving to `Codable`/`Decodable` in time too. I still have a lot of other work to do in this area too due to a codebase with its roots in swift2.

@jseibert great work with async and the swiftNIO integration. I'm desperate to find the time to look into that stuff but haven't so far.
I've gone pretty far building my services on [perfect.org](perfect.org) rather than vapor so as @paulofaria suggests taking core swiftNIO directly would be ideal.

**Comment by jseibert** [comment]
@sportlabsMike upstream Runtime is merged!

**Comment by sportlabsMike** [comment]
Great! I've raised https://github.com/GraphQLSwift/GraphQL/pull/38
Pointed to Runtime 0.5.0 and cherry picked the fixes to the initialiser from @jseibert 
Thanks everyone.

**Comment by paulofaria** [comment]
Fixed by #38
